### ISSUE TITLE
Add mouse cursor support to the remaining widgets

### ISSUE THREAD

**Issue by dkwingsmt** [flutter/flutter#58192]
While https://github.com/flutter/flutter/pull/57628 added mouse cursor support to a number of material widgets, there are other widgets remaining to support.

- Material:
  - Card (probably)
  - Tooltip
  - NavigationRail
  - ReorderableList
  - Stepper
  - TimePicker
  - Chip
  - DataTable
  - Hyperlinks (not a widget; actually `InlineSpan` with tap gesture recognizers)
- Cupertino widgets (we might not do this until we start to support iPadOS cursors)

Please comment if widgets not mentioned here should also support mouse cursors.

**Comment by happyharis** [comment]
How about Material Card widget?

**Comment by dkwingsmt** [comment]
@happyharis 
1. That's probably a good idea.
2. I don't think so. `GestureDetector` is supposed to be a basic widget that serves one purpose, i.e. define gestures. Developers are supposed to compose it with others if desired. Additionally, `GestureDetector` contains way more than just clicking. For example, it supports dragging, which you might want to use the "drag" cursor instead.
3. It's actually `RichText` with an `InlineSpan` that has recognizers; however there is a "hyperlink_test.dart".

**Comment by gskinner-shared** [comment]
If this is not getting added to gesture detector, is there plans to provide a less opinionated btn than RawMaterialBtn or Inkwell?

Currently they are the next step up from GT, and are filled with material-specifix behavior and complexity that may not be desired.

**Comment by esDotDev** [comment]
If this is not being added to GT, is there plans to provide a less opinionated HitArea?

Currently the next step up from GT is Inkwell and RawMaterialBtn, which are loaded with material specific behavior and issues which be undesired.

A standardized Widget that provides the Event surface of RawMatBtn, with builtin cursor handling would essentially deprecate GT in most code bases where GT is used primarily for taps, and could itself then be used to compose custom buttons. 

I can't think of any cases where someone using GT for taps, would not want cursor support.

**Comment by dkwingsmt** [comment]
@esDotDev By GT do you mean GestureDetector? I don’t quite get you yet, but does MouseRegion, which has a `cursor` property to customize cursors work for you?

**Comment by esDotDev** [comment]
Yes I mean Gesture Detector. 

Well, for use case, consider I need to build any simple non-material button. Ideally I could create this without writing 100+ lines of boilerplate to handle basic Btn functionality.

I would like to have basic button events like:
- onDoubleTap 
- onFocusChange 
- onHighlightChanged
- onHover 
- onLongPress 
- onTap 
- onTapCancel 
- onTapDown 

I would also like to have standard button behavior, like a mouse cursor on roll-over. 

It's not that this would be a hard Widget to build, but the absence of something like this feels wrong. If GestureDetector is truly to be a single purpose widget, then it seems to make sense that the Flutter team would compose it + focus node + mouse region into a simple straightforward "Button" that we can then use as a key building block in all our custom designed.

**Comment by dkwingsmt** [comment]
@esDotDev I suggest raising it as a separate issue, since it sounds highly above the scope of this issue (although personally I don't think it sounds "basic" and widespread enough).

**Comment by esDotDev** [comment]
It's literally applying the same single responsibility principle you just championed for GT, but for a clickable region. 

This seem extremely basic/widespread to me.
```
Color _overColor;
Color _focusColor;

return SimpleButton(
   child: Container(color: _focusColor ?? _overColor ?? Colors.red)
   onHover: (v) => _overColor = v? Colors.green : null,
   onFocusChange : (v) => _focusColor = v? Colors.blue : null,
)
```

**Comment by dkwingsmt** [comment]
@esDotDev Nevertheless, would you mind making it a separate issue? It'll get more notice and discussion in that way.

**Comment by jlubeck** [comment]
@dkwingsmt you mentioned that `It's actually RichText with an InlineSpan that has recognizers;` however, when adding a `TapGestureRecognizer` to that, it does not show the mouse cursor which would be expected

**Comment by dkwingsmt** [comment]
@jlubeck Yes that's intended (although undesired). `TapGestureRecognizer` only allows textspans to respond to pointer events. It takes something else to make it change mouse cursors, however the current design makes it very hard. I'm aware that cursor on links is an important demand and will take it as a high priority issue.

**Comment by jbryanh** [comment]
How about DropdownButtonFormField ?

**Comment by dkwingsmt** [comment]
`TextSpan.mouseCursor` has been added with https://github.com/flutter/flutter/pull/77754#event-4506166822.

**Comment by dancojocaru2000** [comment]
Since macOS is a target, why aren't Cupertino widgets considered until iPadOS is supported?

**Comment by guidezpl** [comment]
For the `GestureDetector` question, I'm doing 

```
return MouseRegion(
  cursor: SystemMouseCursors.click,
  child: GestureDetector(
    onTap: () {
      ...
    },
    child: ...,
  ),
);
```

**Comment by dkwingsmt** [comment]
> Since macOS is a target, why aren't Cupertino widgets considered until iPadOS is supported?

It depends on whether you're talking about desktop cursors or iPadOS cursors. 

The iPadOS cursors are very different from what we currently have and needs non-trivial exploration, and only applies to iPadOS anyway.

For desktop cursors, the question comes down to the design: which parts of Cupertino widgets should we assign cursors to, and what cursors should we use? Since the Cupertino widgets are designed only for iOS, there are no existing guidelines. We can fabricate one, but how do we justify our fabrication?

**Comment by dancojocaru2000** [comment]
Well, Mac Catalyst is a thing so we can just take a native iOS app, put it through Mac Catalyst, and see how macOS shapes the cursor on native iOS controls, then have Cupertino widgets behave the same.

**Comment by dkwingsmt** [comment]
> Well, Mac Catalyst is a thing so we can just take a native iOS app, put it through Mac Catalyst, and see how macOS shapes the cursor on native iOS controls, then have equivalent Cupertino widgets behave the same.

That is a good point. (Although my blind guess is that it will not assign any cursors.)

**Comment by dancojocaru2000** [comment]
From a quick experiment, it seems that it automatically assigns the I-beam cursor for editable text (Flutter seems to do this automatically for `CupertinoTextField` too), and it assigns the hand cursor for links inside labels and text views (and for this I think Flutter requires manually configuring a `TextSpan` to look like a link).

To be fair, on second though, I'm not sure if there is much else that needs a different default cursor. I wonder if the Material widgets have other stuff.

**Comment by esDotDev** [comment]
> For desktop cursors, the question comes down to the design: which parts of Cupertino widgets should we assign cursors to, and what cursors should we use? Since the Cupertino widgets are designed only for iOS, there are no existing guidelines. We can fabricate one, but how do we justify our fabrication?

I think you're overthinking this, no need for guidelines to tell us that text should have "I" cursor and clickable items should have a hand cursor.

**Comment by aytunch** [comment]
> For the `GestureDetector` question, I'm doing
> 
> ```
> return MouseRegion(
>   cursor: SystemMouseCursors.click,
>   child: GestureDetector(
>     onTap: () {
>       ...
>     },
>     child: ...,
>   ),
> );
> ```

So is this the preferred way to show click icon as the cursor for pressable GT widgets in WEB?

**Comment by esDotDev** [comment]
fwiw, I have since discovered the `FocusableActionDetector` which is essentially the 'SimpleButton' API that I was asking for: 
https://blog.gskinner.com/archives/2021/06/flutter-building-custom-components-with-focusableactiondetector.html

With this you can make custom buttons that respond to tab focus, keyboard presses, mouse cursor etc, without having to wrap a bunch of widgets yourself.

**Comment by ValentinVignal** [comment]
@dkwingsmt, @victorsanni, would there be a way for us to identify all the widgets we need (and want) to add a mouse cursor support? For example, for the material ones, maybe it is specified in a material design doc somewhere?


Some of the widgets mentioned in the issue are not trivial. For example, `ReorderableList`, and `TimePicker`. What area of those widgets would we want to apply the given `mouseCursor` parameter?

**Comment by dkwingsmt** [comment]
That's the exact question we're asking and why I hesitate to work on it, especially Cupertino widgets.

`ReorderableList` might be a good way to start, because the dragging gesture isn't immediately obvious and IMO will benefit from a dragging cursor very much. The time picker, on the other hand, might not benefit much and I would suggest be more cautious. 

In general, it would be better to add these features when backed by either specifications, similar widgets in other design widget libraries, or user study (detailed or brief). My personal preference at the moment is to address them when asked by developers with real use case, since otherwise there are probably issues of higher priorities anyway.

**Comment by erf** [comment]
I'm missing a `mouseCursor` field to `PopupMenuButton`.
